File: app/src/main/java/org/sco/movieratings/api/TheMovieDBService.java
package org.sco.movieratings.api;

import org.sco.movieratings.api.response.MoviesResponse;
import org.sco.movieratings.api.response.PreviewsResponse;
import org.sco.movieratings.api.response.ReviewsResponse;

import androidx.annotation.NonNull;
import io.reactivex.Observable;
import retrofit2.http.GET;
import retrofit2.http.Path;

public interface TheMovieDBService {

    @GET("movie/{list}")
    Observable<MoviesResponse> getMovies(@Path("list") @NonNull String list);

    @GET("movie/{id}/videos")
    Observable<PreviewsResponse> getMoviePreviews(@Path("id") @NonNull int movieId);

    @GET("movie/{id}/reviews")
    Observable<ReviewsResponse> getMovieReviews(@Path("id") @NonNull int movieId);

}

File: app/src/main/java/org/sco/movieratings/api/TheMovieDBService.kt
package org.sco.movieratings.api

import org.sco.movieratings.api.response.MoviesResponse
import org.sco.movieratings.api.response.PreviewsResponse
import org.sco.movieratings.api.response.ReviewsResponse
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class TheMovieDBService @Inject constructor(private val movieApi: TheMovieDBApi) {

    suspend fun getPopularMovies(): Result<MoviesResponse> {
        return apiCall { movieApi.getPopularMovies() }
    }

    suspend fun getTopRatedMovies(): Result<MoviesResponse> {
        return apiCall { movieApi.getTopRatedMovies() }
    }

    suspend fun getMoviePreviews(movieId: Int): Result<PreviewsResponse> {
        return apiCall { movieApi.getMoviePreviews(movieId) }
    }

    suspend fun getMovieReviews(movieId: Int): Result<ReviewsResponse> {
        return apiCall { movieApi.getMovieReviews(movieId) }
    }
}